medmodels/data_synthesis/synthesizer.py

```diff
@@ -0,0 +1,132 @@
+"""Abstract class for synthesizers."""
+
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Literal, Optional, Type
+
+import torch
+from torch import nn
+
+from medmodels.data_synthesis.synthesizer_model import SynthesizerModel
+from medmodels.medrecord.medrecord import MedRecord
+
+if TYPE_CHECKING:
+    from medmodels.data_synthesis.builder import SynthesizerBuilder
+
+
+class Synthesizer(nn.Module, metaclass=ABCMeta):
+    """Synthesizer is the abstract class to define synthesizers."""
+
+    _preprocessor: Type[nn.Module]
+    _postprocessor: Type[nn.Module]
+    device: torch.device
+
+    def __init__(
+        self,
+        preprocessor: Type[nn.Module],
+        postprocessor: Type[nn.Module],
+        device: Literal["cpu", "cuda"] = "cuda",
+    ) -> None:
+        """Initializes the synthesizer with the required components.
+
+        Args:
+            preprocessor (Type[nn.Module]): Type of preprocessor module.
+            postprocessor (Type[nn.Module]): Type of postprocessor module.
+            device (Literal["cpu", "cuda"], optional): Device to use for the model. Defaults
+                to "cuda".
+        """
+        super(Synthesizer, self).__init__()
```

Comment:
```suggestion
        super().__init__()
```

should be enough. The other syntax is python 2.0 afaik